### PR TITLE
fix: backward compatibility with version 7.17.0

### PR DESCRIPTION
**Pull request by salvatore-campagna**
PR #83339 was meant to be included in version 7.17.0 but didn't happen.
It landed in version 7.17.1. As a result, version checks in the code are
incorrect. Instead of checking against version 7.17.0 we need to adjust
the code in such a way that it operates correctly with version 7.17.0.
Version 7.17.0 is released and, as as result, it represents the 'reference'
version we would like all following versions to interoperate with.

Unfortunately this means that interoperability between version 7.17.0 and
version 8.0.0 is broken.

In summary this fix needs to be backported to:
* 7.17.1: so to fix interoperability of version 7.17.1 with 7.17.0
* 8.0.1: so to fix interoperability of version 8.0.1 with 7.17.0
* 8.1 (currently BC): so to fix interoperability of version 8.1.0 with 7.17.0
* master (8.2): so to fix interoperability of version 8.2.0 with 7.17.0

Also, when reading and writing the range key, version 7.17.0 checks against
version 6.4.0 which is not possible with the current version in master.

This is a fix for #83533.

### CODE REVIEW

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java [hunk 1]
```diff
@@ -157,6 +157,7 @@ private static String generateKey(double from, double to, DocValueFormat format)
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
+            // NOTE: we would actually need to use Version.V_6_4_0 which is not available, so we use the earliest possible version
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:160]: I don't think 7_17_1 is the earliest possible version though, right?

Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalRange.java:160]: Yes I will remove the comment...I had to change that in one of the following commits...my assumption about the earlier version was incorrect.

**Comment by nik9000** [comment]
I think the version changes look right to me. Do you have to change and BWC test skips in this one too?

**Comment by salvatore-campagna** [comment]
>

**Review by nik9000 — APPROVED** [review]

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/test-full-bwc

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by salvatore-campagna** [comment]
@elasticmachine test this

**Comment by salvatore-campagna** [comment]
@elasticmachine test this please

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by salvatore-campagna** [comment]
I think this is still not enough. Reading the optional double values `originalFrom` and `originalTo` needs to be done only if version is exactly `7.17.0`. `7.17.0` is the only version writing `originalFrom` and `originalTo`.

**Review by nik9000 — APPROVED** [review]

**Comment by salvatore-campagna** [comment]
I have executed backward compatibility tests locally against version `7.17.0`, `7.17.1`, `8.0.0` and `8.1.0` multiple times...I took advantage of randomness to run something like 4 or 5 backward compatibility tests against each version. I didn't see any failure. Also test "Float range" which covers the original issue works fine.

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/full-bwc
